batch/batch/worker/worker.py

```diff
@@ -408,36 +410,38 @@ def __init__(self, job, name, spec):
         self.overlay_path = None
 
         self.image_config = None
+        self.image_digest = None
         self.rootfs_path = None
         scratch = self.spec['scratch']
         self.container_scratch = f'{scratch}/{self.name}'
         self.container_overlay_path = f'{self.container_scratch}/rootfs_overlay'
         self.config_path = f'{self.container_scratch}/config'
 
+        self.container_name = f'batch-{self.job.batch_id}-job-{self.job.job_id}-{self.name}'
+
         self.netns: Optional[NetworkNamespace] = None
         self.process = None
 
-    async def run(self, worker):
+    async def run(self, worker: 'Worker'):
         try:
 
             async def localize_rootfs():
                 async with worker.rootfs_locks[self.image_ref_str]:
                     last_fetched = worker.rootfs_last_fetched.get(self.image_ref_str)
                     five_minutes_ago = time_msecs() - 5 * 60 * 1000
-                    if last_fetched and last_fetched < five_minutes_ago:
-                        assert self.image_ref_str in image_configs
-                        image_digest = get_image_digest(image_configs[self.image_ref_str])
-                        shutil.rmtree(f'/host/rootfs/{image_digest}')
                     if not last_fetched or last_fetched < five_minutes_ago:
-                        worker.rootfs_last_fetched[self.image_ref_str] = time_msecs()
                         await self.pull_image()
+                        worker.rootfs_last_fetched[self.image_ref_str] = time_msecs()
+                        log.info(f'Added image to cache: {self.image_ref_str}')
 
                     self.image_config = image_configs[self.image_ref_str]
-                    image_digest = get_image_digest(self.image_config)
-                    self.rootfs_path = f'/host/rootfs/{image_digest}'
+                    self.image_digest = get_image_digest(self.image_config)
+                    self.rootfs_path = f'/host/rootfs/{self.image_digest}'
```

Comment:
Is `self.image_digest` always a 1:1 mapping with `self.image_ref_str`? It's a bit confusing to me that you have two data structures. One is indexed by image_digest and the other is by image_ref_str.

Reply:
It's not. `image_digest` is the unique SHA for a docker image, while `image_ref_str` is the colloquial name for the image the job will use, i.e. `gcr.io/hail-vdc/batch-worker`. I agree it's confusing to use both, but they serve two different purposes: we need to lock on `image_ref_str` because that's what the jobs are specifying, and we count how many containers are running on a particular image digest because we could have multiple digests downloaded for a particular `image_ref_str`.